### PR TITLE
BTC InputSelector simplification

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Simplification of the Bitcoin InputSelector logic, to work with only amounts, not full UTXO objects.
Triggered by UTXO selection generalization for Avalanche, done as part of #1480 .
Isolated to be included before Avalanche, because it is needed for large number of UTXOs issue, #1639 .

Also include 2 other small changes:
- ConstantFeeCalculator -- very simple for constant fee
- Json comparison in TestUtilities.

## Testing instructions

Unit tests, Bitcoin.

## Types of changes

* Refactor with no change in functionality

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

**Comment by catenocrypt** [comment]
Note: UnspentSelector/InputSelector files has been renamed, in separate commit, then modified, yet diff does not match them all correctly.  Not sure how this could be improved.

File: src/Bitcoin/InputSelector.cpp [hunk 1]
```diff
@@ -92,43 +103,43 @@ UTXOs UnspentSelector::select(const UTXOs& utxos, int64_t targetValue, int64_t b
...
         const auto targetWithFeeAndDust = targetValue + fee + dustThreshold;
         if (maxWithXInputs[numInputs] < targetWithFeeAndDust) {
             // no way to satisfy with only numInputs inputs, skip
             continue;
         }
-        auto slices = slice(sortedUtxos, static_cast<size_t>(numInputs));
+        auto slices = slice(sortedIndices, static_cast<size_t>(numInputs));
+
         slices.erase(
             std::remove_if(slices.begin(), slices.end(),
```
Comment (hewigovens) [inline src/Bitcoin/InputSelector.cpp:116]: let's add a `filter` help method? I wanted to use it too

File: tests/interface/TWTestUtilities.h [hunk 2]
```diff
@@ -28,6 +30,12 @@ inline void assertHexEqual(const std::shared_ptr<TWData>& data, const char* expe
     assertStringsEqual(hex, expected);
 }
 
+inline void assertJSONEqual(std::string& lhs, const char* expected) {
```
Comment (hewigovens) [inline tests/interface/TWTestUtilities.h:33]: this method is already in master

**Review by hewigovens — APPROVED** [review]
so do we need convert to concrete UTXO object to amount and select them by indices, how hard to add a virtual interface?

File: tests/interface/TWTestUtilities.h (hunk 2 above)
Comment (catenocrypt) [reply tests/interface/TWTestUtilities.h]: indeed, refreshed.

**Comment by catenocrypt** [comment]
> so do we need convert to concrete UTXO object to amount and select them by indices, how hard to add a virtual interface?

The alternatives would be to use
1. An interfaces that has an amount()
2. A template type parameter that has an amount

Unfortunatelly even with the virtual interface arrays have to be preprocessed, because vector<UTXO> cannot be cast to vector<InterfaceWithAmount> implicitly, but has to be either array of pointers (discouraged), or some unique_ptr<> (extra complexity).  So the UTXO vector needs some preprocessing before invoking InputSelector anyways.
But the final dereferencing of the returned indices could be simplified if using an interface.

**Comment by catenocrypt** [comment]
> so do we need convert to concrete UTXO object to amount and select them by indices, how hard to add a virtual interface?

I re-did with templates.

File: src/Bitcoin/InputSelector.cpp (hunk 1 above)
Comment (catenocrypt) [reply src/Bitcoin/InputSelector.cpp:120]: Filter what?  Below a limit, or dust? `filterDustInput` and `filterThreshold` is available, I'll make them public.
https://github.com/trustwallet/wallet-core/pull/1650/files#diff-f308f664e0f47afd7ee448c0815cac953b53197567a0b86dcc9fe4de23ab81a8R41

**Review by vikmeup — APPROVED** [review]

**Comment by catenocrypt** [comment]
I will merge this to Avalanche PR #1480 and adapt it.

**Pull request merged**